projects/pgai/pgai/vectorizer/vectorizer.py

```diff
@@ -290,6 +289,102 @@ def fetch_work_query(self) -> sql.Composed:
             source_table=sql.Identifier(self.vectorizer.source_table),
         )
 
+    @cached_property
+    def fetch_work_query_with_retries(self) -> sql.Composed:
+        """
+        Note that this is an updated version of the original query that includes
+        the loading_retries column.
+        Generates the SQL query to fetch work items from the queue table.
+
+        The query is safe to run concurrently from multiple workers. It handles
+        duplicate work items by allowing only one instance of the duplicates to
+        be proccessed at a time.
+
+        For a thorough explanation of the query, see:
+
+        https://www.timescale.com/blog/how-we-designed-a-resilient-vector-embedding-creation-system-for-postgresql-data/#process-the-work-queue
+
+        The main takeaways are:
+
+        > ... the system retrieves a specified number of entries from the work
+        queue, determined by the batch queue size parameter. A FOR UPDATE lock
+        is taken to ensure that concurrently executing scripts don't try
+        processing the same queue items. The SKIP LOCKED directive ensures that
+        if any entry is currently being handled by another script, the system
+        will skip it instead of waiting, avoiding unnecessary delays.
+
+        > Due to the possibility of duplicate entries for the same blog_id
+        within the work-queue table, simply locking said table is insufficient
+        ... A Postgres advisory lock, prefixed with the table identifier to
+        avoid potential overlaps with other such locks, is employed. The try
+        variant, analogous to the earlier application of SKIP LOCKED, ensures
+        the system avoids waiting on locks. The inclusion of the ORDER BY
+        blog_id clause helps prevent potential deadlocks...
+
+        The only difference, between the blog and this query, is that we handle
+        composite primary keys.
+        """
+        return sql.SQL("""
+                WITH selected_rows AS (
+                    SELECT {pk_fields}, {loading_retries}
+                    FROM {queue_table}
+                    WHERE loading_retry_after is null or loading_retry_after < now()
+                    LIMIT %s
+                    FOR UPDATE SKIP LOCKED
+                ),
+                locked_items AS (
+                    SELECT
+                        {pk_fields}, {loading_retries},
+                        pg_try_advisory_xact_lock(
+                            %s,
+                            hashtext(concat_ws('|', {lock_fields}))
+                        ) AS locked
+                    FROM (
+                        SELECT DISTINCT {pk_fields}, {loading_retries}
+                        FROM selected_rows
+                        ORDER BY {pk_fields}
+                    ) as ids
+                ),
+                deleted_rows AS (
+                    DELETE FROM {queue_table} AS w
```

Comment:
I think you actually want the loading_retries from here. Because you want the minimum loading_retries form the rows you delete for any pk. I.e. if you had a loading_retries of 4, but then you update the row and get a new queue entry, you want to restart the loading_retries at 0

Reply:
Could you please elaborate? I'm not sure I understood this comment.

	Reply...

Reply:
Could you please elaborate? I'm not sure I understood this comment

Reply:
Isn't it this query removing the records on every run?
If we get the retries here wouldn't it be 0 always?